docs/developer/benchmark.ipynb

```diff
@@ -0,0 +1,587 @@
+   "source": [
+    "import sys, os\n",
+    "sys.path.insert(0, os.path.abspath('../../'))  # To use ``tests`` as a package."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "db5d1f7c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from rich.pretty import Pretty\n",
+    "from tests.benchmarks.runner import create_benchmark_runner_factory, BenchmarkSession\n",
+    "\n",
+    "benchmark_factory = create_benchmark_runner_factory()\n",
+    "session = benchmark_factory[BenchmarkSession]\n",
+    "Pretty(session, max_depth=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0d75c9c3",
+   "metadata": {},
+   "source": [
+    "Here is the simple use case of ``BenchmarkSession.run``."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bc9eaaa0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def test_func(x: float) -> float:\n",
+    "    from time import sleep\n",
+    "    sleep(x)\n",
+    "\n",
+    "    return x\n",
+    "\n",
+    "session.run(test_func, 0.1)\n",
+    "Pretty(session.report.measurements)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b63a66a",
+   "metadata": {},
+   "source": [
+    "The ``BenchmarkSession.run`` passes all arguments to ``runner`` and append its result into the ``report``.\n",
+    "\n",
+    "You can use ``BenchmarkSession.configure`` to temporarily update configurations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0e056929",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with session.configure(iterations=2):  # Run the benchmark twice.\n",
+    "    session.run(test_func, 0.2)\n",
+    "\n",
+    "Pretty(session.report.measurements)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "096d9fb3",
+   "metadata": {},
+   "source": [
+    "Note that each iteration appends the result separately, instead of deriving average results and save the number of iteration together.\n",
+    "\n",
+    "It is because these tools are intended for time-consuming tests, in the scale of minutes and hours.\n",
+    "\n",
+    "If you need multi-thousands iterations similar to ``timeit``, you can write a special ``runner`` to do so.\n",
+    "\n",
+    "See [Exercies: TimeIt Runner](Exercise:-TimeIt-Runner) for an example."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4ffae990",
+   "metadata": {},
+   "source": [
+    "### BenchmarkRunner\n",
+    "\n",
+    "The ``BenchmarkRunner`` should be a callable that returns a ``SingleRunReport``,\n",
+    "that can be appended to the ``BenchmarkReport``.\n",
+    "\n",
+    "Here is the simple use case of the runner."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "26671cfc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from tests.benchmarks.runner import BenchmarkRunner\n",
+    "runner = benchmark_factory[BenchmarkRunner]  # SimpleRunner\n",
```

Comment:
So `benchmark_factory` is not a factory for benchmarks, but a factory for sessions and runners? Why is it the same factory? What does it do?

Reply:
No, sorry for confusing names... This factory builds a benchmark session object not the benchmarks.
There is another factory to build a prototype.
I renamed it to have clear meaning.